analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/CallModelStateIsValid.cs

```diff
@@ -0,0 +1,117 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Common.Walkers;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class CallModelStateIsValid : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6967";
+    private const string MessageFormat = "ModelState.IsValid should be checked in controller actions.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStart =>
+        {
+            if (compilationStart.Compilation.ReferencesControllers()
+                && !compilationStart.Compilation.Assembly.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ApiControllerAttribute))
+            {
+                compilationStart.RegisterSymbolStartAction(symbolStart =>
+                {
+                    var type = (INamedTypeSymbol)symbolStart.Symbol;
+                    if (type.DerivesFrom(KnownType.Microsoft_AspNetCore_Mvc_ControllerBase)
+                        && !type.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ApiControllerAttribute)
+                        && !HasActionFilterAttribute(type))
+                    {
+                        symbolStart.RegisterSyntaxNodeAction(ProcessControllerMethods, SyntaxKind.MethodDeclaration);
+                    }
+                }, SymbolKind.NamedType);
+            }
+        });
+
+    private static void ProcessControllerMethods(SonarSyntaxNodeReportingContext context)
+    {
+        var method = (MethodDeclarationSyntax)context.Node;
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is IMethodSymbol methodSymbol
+            && methodSymbol.IsControllerMethod()
+            && !HasActionFilterAttribute(methodSymbol)
+            && methodSymbol.Parameters.Any(x => HasValidationAttribute(x) || IsValidatable(x.Type)))
+        {
+            var walker = new ModelStateFinder(context.SemanticModel);
+            walker.Visit(method.Body);
+            if (!walker.ValidatesModel)
+            {
+                context.ReportIssue(Rule, method.Identifier);
+            }
+        }
+    }
+
+    private static bool IsValidatable(ITypeSymbol type) => type != null
```

Comment:
Why do we need to check this? I think every model is validated no matter if it has any validation attributes/interfaces. E.g.:

A model like this
```cs
public class Model
{
    public int Id { get; set; }
}
```
with input like this:
```json
{
  "Id": "Boom"
}
```
should set the ModelState.IsValid to false (Needs confirmation).
See e.g. here https://stackoverflow.com/a/36894368 (might be related to ASP.NET 4.x, though)

Reply:
That's a good point, and it makes the rule so much easier! I'll update the RSPEC.

Reply:
Did you validate/confirm this behavior in a real Asp.Net application? Is `ModelState.IsValid` set or is the pipeline returning an `500 Bad request` before the method is even called?